baseapp/baseapp.go

```diff
@@ -769,6 +769,52 @@ func makeABCIData(msgResponses []*codectypes.Any) ([]byte, error) {
 	return proto.Marshal(&sdk.TxMsgData{MsgResponses: msgResponses})
 }
 
+// enableFraudProofGenerationMode rolls back an app's state to a previous
+// state and enables tracing for the list of store keys
+// It returns the tracing-enabled app along with the trace buffers used
+func (app *BaseApp) enableFraudProofGenerationMode(storeKeys []types.StoreKey, routerOpts map[string]AppOptionFunc) (*BaseApp, map[string]*bytes.Buffer, error) {
+	cms := app.cms.(*multi.Store)
+	lastVersion := cms.LastCommitID().Version
+	previousCMS, err := cms.GetVersion(lastVersion)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Add options for tracing
+	storeTraceBuf := &bytes.Buffer{}
+
+	storeKeyToSubstoreTraceBuf := make(map[string]*bytes.Buffer)
+
+	// Initialize params from previousCMS
+	storeToLoadFrom := make(map[string]types.KVStore)
+	storeKeyNames := make([]string, 0, len(storeKeys))
+	for _, storeKey := range storeKeys {
+		storeKeyName := storeKey.Name()
+		storeKeyNames = append(storeKeyNames, storeKeyName)
+		storeToLoadFrom[storeKeyName] = previousCMS.GetKVStore(storeKey)
+		storeKeyToSubstoreTraceBuf[storeKeyName] = &bytes.Buffer{}
+	}
+
+	// BaseApp, B1
+	options := []AppOption{
+		SetSubstoreTracer(storeTraceBuf),
+	}
+
+	for _, storeKey := range storeKeys {
+		options = append(options, SetTracerFor(storeKey.Name(), storeKeyToSubstoreTraceBuf[storeKey.Name()]))
+		options = append(options, AppOptionFunc(routerOpts[storeKey.Name()]))
+	}
+	newApp, err := SetupBaseAppFromParams(app.name+"WithTracing", app.logger, dbm.NewMemDB(), app.txDecoder, storeKeyNames, app.LastBlockHeight()+1, storeToLoadFrom, options...)
```

Comment:
I would prefer a separate variable like `newBlockHeight = app.LastBlockHeight()+1` at the top of this method and then using it here.